tests/Unit/Utilities/Test_Functional.cpp

```diff
@@ -26,57 +201,169 @@ void test_get_argument() {
+  const Bound small{{0.1, 5.0}};
+
+  const auto generic_unaries =
+      std::make_tuple(std::make_tuple(TestFuncEvalAbs{}, generic),
+                      std::make_tuple(TestFuncEvalAcos{}, mone_one),
+                      std::make_tuple(TestFuncEvalAcosh{}, gt_one),
+                      std::make_tuple(TestFuncEvalAsin{}, mone_one),
+                      std::make_tuple(TestFuncEvalAsinh{}, generic),
+                      std::make_tuple(TestFuncEvalAtan{}, generic),
+                      std::make_tuple(TestFuncEvalAtanh{}, mone_one),
+                      std::make_tuple(TestFuncEvalCos{}, generic),
+                      std::make_tuple(TestFuncEvalCosh{}, generic),
+                      std::make_tuple(TestFuncEvalExp{}, generic),
+                      std::make_tuple(TestFuncEvalImag{}, generic),
+                      std::make_tuple(TestFuncEvalInvSqrt{}, positive),
+                      std::make_tuple(TestFuncEvalLog{}, positive),
+                      std::make_tuple(TestFuncEvalReal{}, generic),
+                      std::make_tuple(TestFuncEvalSin{}, generic),
+                      std::make_tuple(TestFuncEvalSinh{}, generic),
+                      std::make_tuple(TestFuncEvalSqrt{}, positive),
+                      std::make_tuple(TestFuncEvalSquare{}, generic),
+                      std::make_tuple(TestFuncEvalTan{}, generic),
+                      std::make_tuple(TestFuncEvalTanh{}, generic));
+
+  tmpl::for_each<AllTypeList>([&generator, &generic_unaries ](auto x) noexcept {
+    using DistType =
+        typename tt::get_fundamental_type_t<typename decltype(x)::type>;
+    tuple_fold(generic_unaries, [&generator, &x ](auto tup) noexcept {
+      std::get<0>(tup)(
+          generator, UniformCustomDistribution<DistType>{std::get<Bound>(tup)},
+          x);
+    });
+  });
+
+  const auto floating_unaries =
+      std::make_tuple(std::make_tuple(TestFuncEvalLog10{}, positive));
+
+  const auto floating_binaries =
+      std::make_tuple(std::make_tuple(TestFuncEvalPow{}, small));
+
+  const auto generic_binaries =
+      std::make_tuple(std::make_tuple(TestFuncEvalDivides{}, gt_one),
+                      std::make_tuple(TestFuncEvalMinus{}, generic),
+                      std::make_tuple(TestFuncEvalMultiplies{}, generic),
+                      std::make_tuple(TestFuncEvalPlus{}, generic),
+                      std::make_tuple(TestFuncEvalDivAssign{}, gt_one),
+                      std::make_tuple(TestFuncEvalMultAssign{}, generic),
+                      std::make_tuple(TestFuncEvalMinusAssign{}, generic),
+                      std::make_tuple(TestFuncEvalPlusAssign{}, generic));
+
+  tmpl::for_each<DoubleSet>([
+    &generator, &generic_binaries, &floating_unaries, &floating_binaries
+  ](auto x) noexcept {
+    using DistType1 =
+        typename tt::get_fundamental_type_t<typename decltype(x)::type>;
+    tuple_fold(floating_unaries, [&generator, &x ](auto tup) noexcept {
```

Comment:
`auto` -> `const auto&` (no point in making a copy each time)

Reply:
happy to add reference, but const causes problems:
it means the call operator in the macro-generated functions have to be const, which means the distributions passed in have to be const, which causes compiler errors at the point of random number generation because std::uniform...distributions(generator) are not const.

Reply:
oops, nope - references cause the same problem because `tuple_fold_impl` consts its tuples too; so, I either need to make copies or change the way the functions work. For these tests, the copies are pretty small (empty function object and two doubles) - so I think it should be fine. I guess it is conceivable that having a tuple_fold that doesn't const the tuple is desirable, but seems like a discussion beyond this PR.